File: documents/assignment.m
%% Matrix Decompositions in Data Analysis
% Winter 2019
% Assignment file
% FILL IN THE FOLLOWING INFORMATION
% Name: 
% Student ID:
%

%% Main function
% In Matlab, we can't mix script files and function files, so we'll have 
% a main function called assignment that will call the other functions
% that can be added to this file.
%
% For prototyping, it's better to call the functions from Matlab command line, 
% as the functions don't save their working space. 

function assignment() 
  % To read the data, we can use the readtable command
  T = readtable('news.csv');
  A = T.Variables;
  terms = T.Properties.VariableNames;

  %% Task 1
  % FILL IN HERE the code for doing Task 1
  % One round of nmf_als would be like
  [W, H, errs] = nmf_als(A, 20);
  % To plot the errors per iterations, use
  figure, plot(errs);
  title('Convergence of NMF ALS');
  xlabel('Iteration');
  ylabel('||A - WH||_F^2');
  % DO THE OTHER NMF methods similarly and add here code to call them
  % and to do the comparisons. 

  %% Task 2
  % We use B for normalised A
  B = A./sum(sum(A));
  % DO THE NMF

  % If H is an output of some NMF algorihtm, we get the top-10 entries
  % of the first row of H as follows
  h = H(1,:);
  [~, I] = sort(h, 'descend');
  for i=1:10
    fprintf('%s\t%f\n', terms{I(i)}, h(I(i)));
  end

  %% Task 3
  % To compute K-L, we need the z-scores
  Z = normalize(A); % This requires MATLAB 2018a
  [U, S, V] = svds(Z, 20);
  KL = Z*V; % N.B. svds returns correct-sized V
  %% COMPUTE pLSA
  % To compute k-means, use kmeans; Replicates controls the number of 
  % restarts. 
  [idx] = kmeans(KL, 20, 'Replicates', 20);
  % DO K-MEANS for other cases 
  % To calculate the NMI, use
  nmi_KL = nmi_news(idx);
  fprintf('NMI (KL) = %f\n', nmi_KL);
  % DO NMI for other clusterings
end


function [W, H, errs] = nmf_als(A, k, maxiter) 
% Boilerplate function for NMF with ALS
% Errs contains the errors per iteration
% Use 300 iterations if no iters is given
  if nargin < 3 
    maxiter = 300; 
  end
  % Take dimensions and init W
  [n, m] = size(A);
  W = rand(n, k);
  H = rand(k, m); % This is actually not needed
  % Init errs to NaNs
  errs = nan(maxiter, 1);
  for i = 1:maxiter,
    % FILL IN an update for H and W
    errs(i) = norm(A - W*H, 'fro')^2; % Squared Frobenius
  end

end


function z = nmi(x, y)
% Compute normalized mutual information I(x,y)/sqrt(H(x)*H(y)) of two discrete variables x and y.
% Input:
%   x, y: two integer vector of the same length 
% Ouput:
%   z: normalized mutual information z=I(x,y)/sqrt(H(x)*H(y))
% Written by Mo Chen (sth4nth@gmail.com).
% https://de.mathworks.com/matlabcentral/fileexchange/29047-normalized-mutual-information
  assert(numel(x) == numel(y));
  n = numel(x);
  x = reshape(x,1,n);
  y = reshape(y,1,n);
  l = min(min(x),min(y));
  x = x-l+1;
  y = y-l+1;
  k = max(max(x),max(y));
  idx = 1:n;
  Mx = sparse(idx,x,1,n,k,n);
  My = sparse(idx,y,1,n,k,n);
  Pxy = nonzeros(Mx'*My/n); %joint distribution of x and y
  Hxy = -dot(Pxy,log2(Pxy));
  % hacking, to elimative the 0log0 issue
  Px = nonzeros(mean(Mx,1));
  Py = nonzeros(mean(My,1));
  % entropy of Py and Px
  Hx = -dot(Px,log2(Px));
  Hy = -dot(Py,log2(Py));
  % mutual information
  MI = Hx + Hy - Hxy;
  % normalized mutual information
  z = sqrt((MI/Hx)*(MI/Hy));
  z = max(0,z);

end

function z = nmi_news(x)
% Computes the NMI between the news ground truth and given clustering
  gt = load('news_ground_truth.txt');
  z = nmi(x, gt);
end 

File: documents/assignment.R
## Matrix Decompositions in Data Analysis
## Winter 2019
## Assignment file
## FILL IN the following information
## Name:
## Student ID:

##
## This file contains stubs of code to help you to do your 
## assignment. You can fill your parts at the indicated positions
## and return this file as a part of your solution. 
##
## Remember to fill your name and student ID number above.
##


## Task 1
##########

## Boilerplate for ALS NMF
nmf.als <- function(A, k, maxiter=300) {
	result <- NA
	# Initialize W and H (only one is actually needed)
	W <- matrix(runif(nrow(A)*k), nrow(A), k)
	H <- matrix(runif(k*ncol(A)), k, ncol(A))
	errs <- rep(NA, times=maxiter) # Stores errors per iteration
	for (i in 1:maxiter) {
		# FILL IN update for W and H
		errs[i] <- norm(A - W%*%H, 'F')
	}
	# Copy the column names to H; helps analysing it
	colnames(H) <- colnames(A)
	result <- list(W=W, H=H, errs=errs)
	result
}

## Load the news data
A <- as.matrix(read.csv("news.csv"))

## Sample use of nmf.als:
res <- nmf.als(A, 20) # Computes NMF with ALS updates
## To show the per-iteration error
plot(res$errs, type="l", main="Convergence of NMF ALS", xlab="Iteration", ylab="Squared Frobenius")


## IMPLEMENT the other algorithms
## DO the comparisons


#############
## Task 2
#############

## Remember to normalize the data before applying the NMF algorithms
B = A/sum(A)
## If res <- nmf.als(A, 20, ...), you can print the top-10 terms by
for (k in 1:nrow(res$H)) print(sort(res$H[k,], decreasing=TRUE)[1:10])

## USE NMF to analyse the data
## REPEAT the analysis with GKL-optimizing NMF

#############
## Task 3
#############

## General code to compute the NMI
nmi <- function(x, opt) {
  p.x <- table(x)/length(x)
  p.y <- table(opt)/length(opt)
  if (length(p.x) != length(p.y)) {
    stop("Data has incorrect number of clusters")
  }
  p.xy <- table(x, opt)/length(x)

  H.xgy <- 0
  H.ygx <- 0
  H.xy <- 0
  for (i in 1:length(p.y)) {
    for (j in 1:length(p.y)) {
      if (p.xy[i,j] > 0) {
        H.xgy <- H.xgy + p.xy[i,j]*log(p.x[i]/p.xy[i,j])
        H.xy <- H.xy - p.xy[i,j]*log(p.xy[i,j])
      }
      if (p.xy[j,i] > 0) {
        H.ygx <- H.ygx + p.xy[j,i]*log(p.y[j]/p.xy[j,i])
      }
    }
  }
  I <- H.xy - H.xgy - H.ygx
  as.vector((H.xy - I)/H.xy)
}
  
## Code to compute the NMI for the news
nmi.news <- function(x) { nmi(x, as.matrix(read.csv("news_ground_truth.txt",header = FALSE))) }

## Here we compute Karhunen--Loeve
Z <- scale(A) # Z-scores
SVD <- svd(A, nu=0, nv=20)
V = SVD$v
KL = Z%*%V

## COMPUTE pLSA with the matrix B from the previous task

## Do some clustering, here for the KL matrix
clust <- kmeans(KL, 20, iter.max=100, nstart=20)$cluster

## How good is this? The smaller the better
nmi.news(clust)

## DO NMF with GKL optimization and other comparisons
